### PR TITLE
Add strictFuncs to stdlib string libraries

### PR DESCRIPTION
**Pull request by jamesaorson**
Local testing of the string test files showed it passing with the simple addition of the experimental strictFuncs pragma.

### CODE REVIEW

**Comment by jamesaorson** [comment]
Attempting to rebuild nim gives this error
```
bin/nim_csources_a8a5241f9475099c823cfe1a5e0ca4022ac201ff exists.

cp bin/nim_csources_a8a5241f9475099c823cfe1a5e0ca4022ac201ff bin/nim

bin/nim_csources_a8a5241f9475099c823cfe1a5e0ca4022ac201ff -v
Nim Compiler Version 1.0.11 [Linux: amd64]
Compiled at 2020-12-21
Copyright (c) 2006-2019 by Andreas Rumpf

git hash: 19440baa807bbda58290ac9d491c9aa8a2bea2fa
active boot switches: -d:release -d:danger

bin/nim c --skipUserCfg --skipParentCfg koch
Hint: used config file '/mnt/c/Nim/config/nim.cfg' [Conf]
Hint: used config file '/mnt/c/Nim/koch.nim.cfg' [Conf]
Hint: used config file '/mnt/c/Nim/config/config.nims' [Conf]
Hint: system [Processing]
Hint: widestrs [Processing]
Hint: io [Processing]
Hint: koch [Processing]
Hint: os [Processing]
Hint: strutils [Processing]
/mnt/c/Nim/lib/pure/strutils.nim(10, 17) Error: unknown experimental feature
```

Why would it try to use such an old version of nim to build?

If so, how are we expected to be able to add strictFuncs with the feature not existing in the compiler version being used?

**Comment by ghost** [comment]
@JamesOrson because for bootstrapping the Nim compiler uses an older version of itself, right now it's 1.0. https://github.com/nim-lang/csources_v1

**Comment by ghost** [comment]
I don't think you should add this pragma to each module additionally - when people compile with strictFuncs, it'll be applied to all code regardless if it's stdlib or not. But just for reference - you can do something like https://github.com/Yardanico/nimpylib/blob/master/src/pylib.nim#L1 , that's how different features are disabled/enabled based on compiler versions/flags.

**Comment by ringabout** [comment]
please use latest devel nim to build Nim. btw you don't need to build Nim at all when changing stdlib. Just use `nim c -r --lib:lib app.nim`, `--lib:lib` will use the modified stdlib.

**Comment by ghost** [comment]
@xflywind "please use latest devel nim to build Nim" no, that wasn't the issue here at all. The issue here is in bootstrapping, which uses Nim 1.0 specifically.

**Comment by jamesaorson** [comment]
> I don't think you should add this pragma to each module additionally - when people compile with strictFuncs, it'll be applied to all code regardless if it's stdlib or not. But just for reference - you can do something like https://github.com/Yardanico/nimpylib/blob/master/src/pylib.nim#L1 , that's how different features are disabled/enabled based on compiler versions/flags.

The flag check is very helpful, thank you!

**Comment by jamesaorson** [comment]
> I don't think you should add this pragma to each module additionally - when people compile with strictFuncs, it'll be applied to all code regardless if it's stdlib or not. But just for reference - you can do something like https://github.com/Yardanico/nimpylib/blob/master/src/pylib.nim#L1 , that's how different features are disabled/enabled based on compiler versions/flags.

Why would we not want to define this in the stdlib where possible? It only makes sure that we maintain that contract and allows us to solidify "What `func`s are truly without side effects?" within the stdlib.

I am working on a functional game engine in Nim and I want to be able to know what functions from the stdlib I actually CAN use in my pure `func`s.

**Comment by ghost** [comment]
@JamesOrson did you see https://github.com/nim-lang/Nim/blob/devel/tests/effects/tstrict_funcs_imports.nim then?

**Comment by jamesaorson** [comment]
> @JamesOrson did you see https://github.com/nim-lang/Nim/blob/devel/tests/effects/tstrict_funcs_imports.nim then?

I did not! This is really great to see. So this work was started by me seeing issue #16832, which mentioned it would be better to see these all labeled as `func` instead of `proc` where we can. So really I should focus on renaming `proc` to `func` wherever it is valuable, for the sake of documentation. Would you agree?

File: lib/pure/strformat.nim [hunk 1]
```diff
@@ -316,16 +316,16 @@ single letter DSLs.
...
   if v < 10:
     result = $chr(ord('0') + v)
   else:
     result = $chr(ord(if typ == 'x': 'a' else: 'A') + v - 10)
 
-proc alignString*(s: string, minimumWidth: int; align = '\0'; fill = ' '): string =
+func alignString*(s: string, minimumWidth: int; align = '\0'; fill = ' '): string =
   ## Aligns `s` using the `fill` char.
-  ## This is only of interest if you want to write a custom `format` proc that
+  ## This is only of interest if you want to write a custom `format` func that
```
Comment (konsumlamm) [inline lib/pure/strformat.nim]:
```suggestion
  ## This is only of interest if you want to write a custom `format` proc that
```
Here it makes sense to say proc, this is not a definition, ditto for the other cases in doc comments (except for links).

**Review by konsumlamm — CHANGES_REQUESTED** [review]

File: lib/pure/strformat.nim (hunk 1 above)
Comment (jamesaorson) [reply lib/pure/strformat.nim]: Done

**Pull request closed without merging**